feat: export token expired error

crenshaw-dev opened the pull request:
Relates to #327 

This allows `errors.Is` checking while preserving the exact same error messages for those who are doing string comparison.

Preserving the exact same strings means some of the error messages (without context) look a little weird. But I think it's a reasonable tradeoff until it's time for another major version bump.

This PR would let me write unit tests in Argo CD without worrying that a change in go-oidc will break my string comparison checks.

ericchiang commented:
Hey @crenshaw-dev,

I don't want to export every possible error. If you've got a specific use case where a typed error would be helpful (e.g. specific HTTP status codes from the provider), happy to add that. Do you have an example of what you're trying to do?

Also, some of these errors arguably shouldn't be used problematically, since they're hit before we validate the signature check on the ID token. 

Eric

crenshaw-dev commented:
@ericchiang there are two current use cases in Argo CD:

1) We want to know if it's a "token expired" error so the UI can redirect the user immediately to a new login flow: https://github.com/argoproj/argo-cd/blob/a588e63e3cb9db91fd491c0d7371c87867861aa7/util/session/sessionmanager.go#L493-L504
2) We have a regression test to make sure we support algorithms besides RS256 (added support with 2.2.0): https://github.com/argoproj/argo-cd/blob/a588e63e3cb9db91fd491c0d7371c87867861aa7/util/session/sessionmanager_test.go#L581

It's possible that test shouldn't exist. But I think the expired example is a reasonable use case. Would you be interested in a PR that only exports that error?

ericchiang commented:
I think a token expired error would be reasonable to add!

That second test probably shouldn't be necessary? Just make sure your RS512 token doesn't return an error.

So sounds like the API we'd want to add it

```
var ErrTokenExpired = errors.New(...)
```

(nit: exported error variables generally start with "Err", and error types end with "Error" https://pkg.go.dev/io#pkg-variables, https://pkg.go.dev/net#AddrError)

Kinda a tangent, but I've also been thinking about having an oidctest package that would let you stand up a httptest provider with some basic configuration. Maybe that would be helpful too.

crenshaw-dev commented:
> Just make sure your RS512 token doesn't return an error.

Yeeeeah that would be proper. I was lazy and didn't want to have to set up a full mock OIDC provider. Maybe that would be easier than I assumed.

> (nit: exported error variables generally start with "Err", and error types end with "Error" https://pkg.go.dev/io#pkg-variables, https://pkg.go.dev/net#AddrError)

Good catch, will fix!

> Kinda a tangent, but I've also been thinking about having an oidctest package that would let you stand up a httptest provider with some basic configuration. Maybe that would be helpful too.

Yep, that would make some other tests I'm working on a lot easier I think! Maybe I can put some Intuit time towards that eventually.

Updating the PR to be more minimal!

crenshaw-dev committed:
feat: export token expiration error

try to match style

follow naming convention

tests

revert import change

match format

naming convention

test

errors.As instead of errors.Is

ericchiang commented on oidc/verify.go:
```diff
@@ -13,14 +13,19 @@ import (
 	"time"
 
 	"golang.org/x/oauth2"
-	jose "gopkg.in/square/go-jose.v2"
```
nit: please leave this line alone. I think adding this is the default goimports functionality.

crenshaw-dev replied:
Sure thing. Gonna fight with GoLand real quick.  :-)

ericchiang commented on oidc/verify.go:
```diff
@@ -13,14 +13,19 @@ import (
+// TokenExpiredErr indicates that Verify failed because the token was expired. This
+// error does NOT indicate that the token is not also invalid for other reasons. Other
+// checks might have failed if the expiration check had not failed.
+var TokenExpiredErr = errors.New("oidc: token is expired")
```
nit: This should be ErrTokenExpired

Or maybe you want to include the expiry?

```
type TokenExpiredError struct {
    Expiry time.Time
}
```

crenshaw-dev replied:
I like that

ericchiang reviewed (commented):
(please squash your commits)

crenshaw-dev commented on oidc/verify_test.go:
```diff
@@ -12,6 +13,9 @@ import (
 )
 
+// anyError is a fake error to match any error returned in a test.
+var anyError = errors.New("")
```
This may or may not be the proper way to do this. lmk

ericchiang replied:
Add a new field and use that instead?

```
type verificationTest struct {
	wantErr bool
	
	wantErrIs error
}
```

crenshaw-dev commented:
@ericchiang ready for another look

ericchiang commented on oidc/verify.go:
```diff
@@ -21,6 +21,25 @@ const (
+
+// Is returns true if the error was due to an expired token. It does not check that
+// the expiry time is identical.
+func (e *TokenExpiredError) Is(target error) bool {
```
Implementing is shouldn't be necessary. Can you drop this? https://go.dev/play/p/fTGu3-8H819

crenshaw-dev replied:
Apparently `errors.Is` is doing string comparison. Since the Expiry is included in the time, I have to use a custom `Is` to avoid checking the Expiry.

crenshaw-dev replied:
Demo: https://go.dev/play/p/7rN5_y9ODKD

ericchiang replied:
You use errors.As instead for error types, not errors.Is:

```
var expErr *oidc.TokenExpiryError
if errors.As(err, &expErr) {
    // ...
}
```

Would recommend dropping Is() implementation. That's generally reserved for logic that needs to be very custom.

crenshaw-dev commented:
@ericchiang updated

ericchiang commented:
thanks!
